File: sources/academy/webscraping/scraping_basics_javascript/exercises/unesco_whs_counts.mjs
import * as cheerio from 'cheerio';

async function download(url) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
  const html = await response.text();
  return cheerio.load(html);
}

function parseWhsCount($) {
  for (const element of $('.card-body').toArray()) {
    const $infoCard = $(element);
    const title = $infoCard.find('.card-title').text();

    if (title.includes('World Heritage Sites')) {
      const number = $infoCard.find('.card-number').text().trim();
      return Number.parseInt(number, 10);
    }
  }
  return 0;
}

const listingUrl = 'https://www.unesco.org/en/countries';
const $listing = await download(listingUrl);
const countries = $listing('.node--type-country').toArray();

for (const element of countries.slice(0, 10)) {
  const $countryCard = $listing(element);
  const $link = $countryCard.find('a').first();
  const href = $link.attr('href');

  if (!href) {
    continue;
  }

  const countryUrl = new URL(href, listingUrl).href;
  const $country = await download(countryUrl);
  const whsCount = parseWhsCount($country);
  console.log(`${countryUrl} ${whsCount}`);
}

File: sources/academy/webscraping/scraping_basics_python/exercises/unesco_whs_counts.py
import httpx
from bs4 import BeautifulSoup
from urllib.parse import urljoin


def download(url: str) -> BeautifulSoup:
    response = httpx.get(url)
    response.raise_for_status()
    return BeautifulSoup(response.text, "html.parser")


def parse_whc_count(soup: BeautifulSoup) -> int:
    for info_card in soup.select(".card-body"):
        title = info_card.select_one(".card-title").text
        if "World Heritage Sites" in title:
            return int(info_card.select_one(".card-number").text.strip())
    return 0


listing_url = "https://www.unesco.org/en/countries"
listing_soup = download(listing_url)
countries = listing_soup.select(".node--type-country")

for country in countries[:10]:
    link = country.select_one("a")
    if not link or 'href' not in link.attrs:
        continue

    country_url = urljoin(listing_url, link["href"])
    country_soup = download(country_url)
    whs_count = parse_whc_count(country_soup)
    print(country_url, whs_count)
